File: codewars/kata-completed/find_parity_outlier/find_parity_outlier.cpp
//#include <print> //C++ 20 only, file not found, as Clang 8 C++17
//#include <string> // not needed now
//#include <iostream> // not needed now
//#include <limits> // not needed now

// function prototype declarations
int GetVectorSize(std::vector<int> arr);

/*
** int s are signed by default so
** the all the integer array items in vector<int>
** are signed int
** therefore in bit wise operations assume singed int
*/
int FindOutlier(std::vector<int> arr)
{
    int result; // the reuturn value
    int arr_size; // the size of the array.
    int even_count; // to test to see if list is even or odd
    
    arr_size = GetVectorSize(arr); // return the number of array elements
    // std::cout << "vector num or elements? is " << arr_size << "\n"; // debug
    
    // instantiate the even_count to 0 or unexpected values are set.
    even_count = 0;
  
    // for the first three numbers, to check list even or odd
    // to set outlier test  
    for (int i = 0; i < 3; i++) {
      int number = arr[i]; 
      // get the first bit, if 1 odd, if 0 even
      // not important to test if its negative or positive
      unsigned int test_bit = 0x1 & number; 
      //std::cout << "first number? is " << number << std::endl; // debug
      //std::cout << "first bit? is " << test_bit << std::endl; // debug
      // the number is even if its first bit is 0
      if (test_bit == 0) {
        even_count++; // increment to count of even numbers by one
      }
    }
  
    //std::cout << "even count? is " << even_count << std::endl; // debug
  
    bool outlier_even = true;
    int iter_limit = 0;
    
    if (even_count == 3) { // the list is even
      // we did not find the odd item, we have go through the entire array until found
      outlier_even = false; // we will will have test for an odd number as outlier
      iter_limit = arr_size;
    } else if (even_count == 2) { // the list is even
      // we did find the odd item, we only have to go through the first three elements
      outlier_even = false; // we will will have test for an odd number as outlier
      iter_limit = 3;
    } else if (even_count == 1) { // the list is odd
      // we did find the even item, we only have to go through the first three elements
      // outlier_even = true; // is already set as the default
      iter_limit = 3;
    } else if (even_count == 0) { // this list is odd
      // we did not find the even item, we have go through the entire array until found
      // outlier_even = true; // is already set as the default
      iter_limit = arr_size;
    }
  
    // the test, finnally
    int i; // to test later for break
    for (i = 0; i < iter_limit; i++) {
      if (outlier_even && ((0x1 & arr[i]) == 0)) {
        result = arr[i];
        break; // we found the item, stop looping
      } else if (!outlier_even && ((0x1 & arr[i]) == 1)) {
        result = arr[i];
        break; // we found the item, stop looping
      }
    }
  
    // break works! i is lower than the arr_size in simple tests
    //std::cout << "i is " << i << ", arr size is " << arr_size << std::endl; // debug
  
    //result = arr_size;
    return result;
}

/*
** return the number of integers items stored in
** the paramater std::vector<int> arr 
**
*/
int GetVectorSize(std::vector<int> arr) {
  
  int result; // the return value
  int arr_size; // the size of the array.
  
  // tried lots of more verbose things before this
  // this simple statement could be moved to main function
  // but seperate function required a function declaration
  // which I had forgotten about, so leaving as is
  arr_size = arr.size();
  result = arr_size; //purposely instantiate the result
  
  return result;
  
}

File: codewars/kata-completed/find_parity_outlier/find_parity_outlier.java
//import java.lang.Math;

public class FindOutlier {
    static int find(int[] integers) {
      
      // get the first three values of a list/array of minimum length of three items
      int firstVal = integers[0];
      int secondVal = integers[1];
      int thirdVal = integers[2];
      int even_count = 0;
      int return_val = 0;
      boolean even_list = false;
      
      // check first bit value in number
      // if the first bit is 0 the number is even
      if ((firstVal & 1) == 0) { 
        //even_count =+ even_count + 1
        even_count ++;
      } 
      
      if ((secondVal & 1) == 0) { 
        //even_count =+ even_count + 1;
        even_count ++;
      }
      // is this one just redundant? probs
       if ((thirdVal & 1) == 0) { 
        //even_count =+ even_count + 1;
        even_count ++;
      } 
      
      /*
      System.out.print("event count " + even_count + "\n"); // debug
      System.out.print("first val " + firstVal + "\n"); // debug
      System.out.print("second val " + secondVal + "\n"); // debug
      System.out.print("thrid val " + thirdVal + "\n"); // debug
      System.out.print("list length " + integers.length + "\n"); // debug
      */
      
      // the number of items needed to iterate over in the list.
      int n = 0;
      
      // list is even and we do have an odd return value to find from the first three items
      if (even_count == 2) {
        even_list = true;
        n = 3;
      // list is odd and we do have an even return value to find from the first three items
      } else if (even_count == 1) {
        even_list = false;
        n = 3;
      // list is odd and we don't have a return value
      } else if (even_count == 0) { // redundant comment out?
       //even_list = false; // redundant already set as default
       //System.out.print("third if else " + even_list + "\n"); // debug, is this ever called, yes!
        n = integers.length;
      } else { // we must itterate across the entire list
        even_list = true;
        n = integers.length;
        //System.out.print("last else " + even_list + "\n"); // debug, is this ever called, yes!
      }
      
      for (int i = 0; i < n; i++) {
        // the list is even check for odd number to return
        if (even_list && ((integers[i] & 1) != 0)) {
          return_val = integers[i];
          
        // the list is odd check for even number to return
        } else if (!even_list && ((integers[i] & 1) == 0)) {
          return_val = integers[i];
        }
      }
      
      // System.out.print("return value " + return_val + "\n"); // debug
      
      return return_val;
    }
  }